resolwe_bio/processes/variant_calling/wgs_preprocess.py

```diff
@@ -90,49 +95,98 @@ class Output:
         sorted_rg = "sorted_rg.bam"
         recal_table = "recal_data.csv"
 
-        # Concatenate multi-lane read files
-        (
-            Cmd["cat"][[reads.path for reads in inputs.reads.output.fastq]]
-            > "input_reads_mate1.fastq.gz"
-        )()
-        (
-            Cmd["cat"][[reads.path for reads in inputs.reads.output.fastq2]]
-            > "input_reads_mate2.fastq.gz"
-        )()
+        index_fasta_name = Path(inputs.bwa_index.output.fasta.path).name
 
-        self.progress(0.05)
-
-        # Align reads with BWA MEM
-        bwa_inputs = [
-            "-K 100000000",
-            "-v 3",
-            f"-t {self.requirements.resources.cores}",
-            "-Y",
-            f"{Path(inputs.bwa_index.output.index.path) / index_fasta_name}",
-            "input_reads_mate1.fastq.gz",
-            "input_reads_mate2.fastq.gz",
-        ]
-        (Cmd["bwa"]["mem"][bwa_inputs] > aligned_sam)()
-        self.progress(0.2)
+        if not inputs.reads and not inputs.aligned_reads:
+            self.error("Please provide FASTQ or BAM input files.")
+        if inputs.reads and inputs.aligned_reads:
+            self.error(
+                "Please provide input data in either FASTQ or aligned BAM format, not both."
+            )
+
+        if inputs.reads:
+            # Define output file names
+            name = inputs.reads.entity_name
+            if not name:
+                mate1_path = Path(inputs.reads.output.fastq[0].path).name
+                assert mate1_path.endswith(".fastq.gz")
+                name = mate1_path[:-9]
+
+            # Concatenate multi-lane read files
+            (
+                Cmd["cat"][[reads.path for reads in inputs.reads.output.fastq]]
+                > "input_reads_mate1.fastq.gz"
+            )()
+            (
+                Cmd["cat"][[reads.path for reads in inputs.reads.output.fastq2]]
+                > "input_reads_mate2.fastq.gz"
+            )()
+
+            self.progress(0.05)
+
+            # Align reads with BWA MEM
+            bwa_inputs = [
+                "-K 100000000",
+                "-v 3",
+                f"-t {self.requirements.resources.cores}",
+                "-Y",
+                f"{Path(inputs.bwa_index.output.index.path) / index_fasta_name}",
+                "input_reads_mate1.fastq.gz",
+                "input_reads_mate2.fastq.gz",
+            ]
+            (Cmd["bwa"]["mem"][bwa_inputs] > aligned_sam)()
+            self.progress(0.2)
+
+        else:
+            # Define output file names
+            name = inputs.aligned_reads.entity_name
+            if not name:
+                bam_path = Path(inputs.aligned_reads.output.bam.path).name
+                assert bam_path.endswith(".bam")
+                name = bam_path[:-4]
+
+            collate_inputs = [
+                f"-@ {self.requirements.resources.cores}",
+                "-O",
+                inputs.aligned_reads.output.bam.path,
```

Comment:
It would be nice to check if species and build are the same for aligned reads and index file.